README.md
# seedwater
Models for Drying and Soaking Kinetics of Seeds

# Bug reports
Please open an issue [here](https://github.com/arsilva87/seedwater/issues).

# Contact
By [e-mail](anderson.silva@ifgoiano.edu.br)


NEWS
------------------------------------------------
                version 2.1

* Starting message fixed

------------------------------------------------
                version 2.0

* Package tkrplot was moved to Suggests and used conditionally

------------------------------------------------
                version 1.0

* First version released on CRAN







R/zz.R
# AUXILIARY FUNCTIONS

# ----------------------------------------------
# Rsq(), aux function
Rsq <- function(model) 
{
    pred <- predict(model)
    n <- length(pred)
    res <- resid(model)
    w <- weights(model)
    if (is.null(w)) 
        w <- rep(1, n)
    rss <- sum(w * res^2)
    resp <- pred + res
    center <- weighted.mean(resp, w)
    r.df <- summary(model)$df[2]
    int.df <- 1
    tss <- sum(w * (resp - center)^2)
    r.sq <- 1 - rss/tss
    adj.r.sq <- 1 - (1 - r.sq) * (n - int.df)/r.df
    out <- list(pseudo.R.squared = r.sq, adj.R.squared = adj.r.sq)
    return(out)
}


# --------------------------------------
# on loading seedwater
.onAttach <- function(lib, pkg)
{
   vers <- read.dcf(file.path(lib, pkg, "DESCRIPTION"), "Version")
   packageStartupMessage(paste("---\nseedwater version", vers))
}

R/soakingmodels.R
soakingmodels <- 
function(moisture, time) 
{
    if (!requireNamespace(c("tcltk", "tkrplot"), quietly = TRUE)) 
        stop("packages tcltk and tkrplot are required")
    geterrmessage()
    done <- tclVar(0)
    soakingModels.env <- new.env()
    assign("soakingModels.tmp", list(), envir = soakingModels.env)
    x <- NULL

    # Water absorption (percentual)
    Mi <- min(moisture)
    Wa <- 100*(moisture - Mi)/Mi

    # equations
    peleg <- function(x, k1, k2) x/(k1 + k2*x)
    logi <- function(x, a, b1, c1) a/(1 + exp(-b1*(x - c1)))
    logi2 <- function(x, a, b1, b2, c1, c2) a/(1 + exp(-b1*(x - c1))) + exp(b2*(x - c2))
    peleg2 <- function(x, k1, k2, b2, c2) x/(k1 + k2*x) + exp(b2*(x - c2))

    # input-start variables
    f <- tclVar("Logistic")
    a <- tclVar("70")
    k1 <- tclVar("0.1")
    k2 <- tclVar("0.9")
    b1 <- tclVar("0.1")
    b2 <- tclVar("0.09")
    c1 <- tclVar("10")
    c2 <- tclVar("50")

    replot <- function(...) {
        fun <- as.character(tclObj(f))
        a <- as.numeric(tclObj(a))
        k1 <- as.numeric(tclObj(k1))
        k2 <- as.numeric(tclObj(k2))
        b1 <- as.numeric(tclObj(b1))
        b2 <- as.numeric(tclObj(b2))
        c1 <- as.numeric(tclObj(c1))
        c2 <- as.numeric(tclObj(c2))

        plot(time, Wa, xlab = "Time", ylab = "Water absorption (%)")
        if (fun == "Peleg") {
            curve(peleg(x, k1, k2), add = TRUE, col = "red")
        } else if (fun == "Logistic") {
            curve(logi(x, a, b1, c1), add = TRUE, col = "red")
        } else if (fun == "Logistic.adapted") {
            curve(logi2(x, a, b1, c1, b2, c2), add = TRUE, col = "red")
        } else if (fun == "Peleg.adapted") {
            curve(peleg2(x, k1, k2, b2, c2), add = TRUE, col = "red")
        }
    }
    redraw <- function(...) {
        fun <- as.character(tclObj(f))
        if (fun == "Peleg") {
            # state of parameters: a, k1, k2, b1, b2, c1, c2
            states <- c("disabled", "normal", "normal", "disabled", "disabled", "disabled", "disabled")
            tkfocus(entry.k1)
            tkfocus(entry.k2)
        } else if (fun == "Logistic") {
            states <- c("normal", "disabled", "disabled", "normal", "disabled", "normal", "disabled")
            tkfocus(entry.a)
            tkfocus(entry.b1)
            tkfocus(entry.c1)
        } else if (fun == "Logistic.adapted") {
            states <- c("normal", "disabled", "disabled", "normal", "normal", "normal", "normal")
            tkfocus(entry.a)
            tkfocus(entry.b1)
            tkfocus(entry.b2)
            tkfocus(entry.c1)
            tkfocus(entry.c2)
        } else if (fun == "Peleg.adapted") {
            states <- c("disabled", "normal", "normal", "disabled", "normal", "disabled", "normal")
            tkfocus(entry.k1)
            tkfocus(entry.k2)
            tkfocus(entry.b2)
            tkfocus(entry.c2)
        }
        tkconfigure(entry.a, state = states[1])
        tkconfigure(ts1, state = states[1]) 
        tkconfigure(entry.k1, state = states[2])
        tkconfigure(ts2, state = states[2]) 
        tkconfigure(entry.k2, state = states[3])
        tkconfigure(ts3, state = states[3])  
        tkconfigure(entry.b1, state = states[4])
        tkconfigure(ts4, state = states[4])  
        tkconfigure(entry.b2, state = states[5])
        tkconfigure(ts5, state = states[5])  
        tkconfigure(entry.c1, state = states[6])
        tkconfigure(ts6, state = states[6])  
        tkconfigure(entry.c2, state = states[7])
        tkconfigure(ts7, state = states[7])  
        replot()
    } 

    base <- tktoplevel()
    tkwm.title(base, "Seed Soaking Kinetics")
    spec.frm <- tkframe(base, borderwidth = 2)
    left.frm <- tkframe(spec.frm)
    right.frm <- tkframe(spec.frm) 

    # parameter: a
    frame1 <- tkframe(left.frm, relief = "groove", borderwidth = 2)
    tkpack(tklabel(frame1, text = "Parameters"), fill = "both", side = "top")
    entry.a <- tkentry(frame1, width = "6", textvariable = a)
    tkpack(ts1 <- tkscale(frame1, label = "a", command = replot, 
        from = 0, to = 100, showvalue = 70, variable = a, 
        resolution = 1, orient = "horiz", relief = "groove"), 
        fill = "both", expand = 1, padx = 3, pady = 2, ipadx = 3, 
        ipady = 2, side = "left")
    tkpack(entry.a, fill = "none", side = "right")

    # models
    frame2 <- tkframe(right.frm, relief = "groove", borderwidth = 2)
    tkpack(tklabel(frame2, text = "Models"))
    for (i in c("Peleg", "Logistic", "Logistic.adapted", "Peleg.adapted")) {
        tmp <- tkradiobutton(frame2, command = redraw, text = i, 
            value = i, variable = f)
        tkpack(tmp, anchor = "w")
    }

    # parameter: k1
    frame3 <- tkframe(left.frm, relief = "groove", borderwidth = 2)
    entry.k1 <- tkentry(frame3, width = "6", textvariable = k1)
    tkpack(ts2 <- tkscale(frame3, label = "k1", command = replot, 
        from = 0, to = 2, showvalue = 0.1, variable = k1,
        resolution = 0.001, orient = "horiz", relief = "groove"), 
        fill = "both", expand = 1, padx = 3, pady = 2, ipadx = 3, 
        ipady = 2, side = "left")
    tkpack(entry.k1, fill = "none", side = "right")

    # parameter: k2
    frame4 <- tkframe(left.frm, relief = "groove", borderwidth = 2)
    entry.k2 <- tkentry(frame4, width = "6", textvariable = k2)
    tkpack(ts3 <- tkscale(frame4, label = "k2", 
        command = replot, from = 0, to = 0.5, showvalue = 0.09, 
        variable = k2, resolution = 0.0001, orient = "horiz", 
        relief = "groove"), fill = "none", expand = 1, padx = 3, 
        pady = 2, ipadx = 3, ipady = 2, side = "left")
    tkpack(entry.k2, side = "right")

    # fitting criteria
    frame5 <- tkframe(right.frm, relief = "groove", borderwidth = 2)
    r2Text <- tclVar("R-sq: ")
    label.r2 <- tklabel(frame5, text=tclvalue(r2Text))
    tkconfigure(label.r2, textvariable=r2Text)
    tkgrid(label.r2)
    r2adjText <- tclVar("Adj. R-sq: ")
    label.r2adj <- tklabel(frame5, text=tclvalue(r2adjText))
    tkconfigure(label.r2adj, textvariable=r2adjText)
    tkgrid(label.r2adj)
    seText <- tclVar("S.E.: ")
    label.se <- tklabel(frame5, text=tclvalue(seText))
    tkconfigure(label.se, textvariable=seText)
    tkgrid(label.se)

    # parameter: b1
    frame6 <- tkframe(left.frm, relief = "groove", borderwidth = 2)
    entry.b1 <- tkentry(frame6, width = "6", textvariable = b1)
    tkpack(ts4 <- tkscale(frame6, label = "b1", 
        command = replot, from = -1, to = 2, showvalue = 0.1, 
        variable = b1, resolution = 0.001, orient = "horiz", 
        relief = "groove"), fill = "none", expand = 1, padx = 3, 
        pady = 2, ipadx = 3, ipady = 2, side = "left")
    tkpack(entry.b1, side = "right")

    # parameter: b2
    frame7 <- tkframe(left.frm, relief = "groove", borderwidth = 2)
    entry.b2 <- tkentry(frame7, width = "6", textvariable = b2)
    tkpack(ts5 <- tkscale(frame7, label = "b2", 
        command = replot, from = -1, to = 2, showvalue = 0.09, 
        variable = b2, resolution = 0.0001, orient = "horiz", 
        relief = "groove"), fill = "none", expand = 1, padx = 3, 
        pady = 2, ipadx = 3, ipady = 2, side = "left")
    tkpack(entry.b2, side = "right")

    # parameter: c1
    frame8 <- tkframe(left.frm, relief = "groove", borderwidth = 2)
    entry.c1 <- tkentry(frame8, width = "6", textvariable = c1)
    tkpack(ts6 <- tkscale(frame8, label = "c1", 
        command = replot, from = 0, to = 200, showvalue = 15, 
        variable = c1, resolution = 1, orient = "horiz", 
        relief = "groove"), fill = "none", expand = 1, padx = 3, 
        pady = 2, ipadx = 3, ipady = 2, side = "left")
    tkpack(entry.c1, side = "right")

    # parameter: c2
    frame9 <- tkframe(left.frm, relief = "groove", borderwidth = 2)
    entry.c2 <- tkentry(frame9, width = "6", textvariable = c2)
    tkpack(ts7 <- tkscale(frame9, label = "c2", 
        command = replot, from = 0, to = 200, showvalue = 50, 
        variable = c2, resolution = 1, orient = "horiz", 
        relief = "groove"), fill = "none", expand = 1, padx = 3, 
        pady = 2, ipadx = 3, ipady = 2, side = "left")
    tkpack(entry.c2, side = "right")

    # buttons
    OnQuit <- function() {
        tclvalue(done) <- 2
    }
    OnClear <- function() {
        assign("soakingModels.tmp", list(), envir = soakingModels.env)
        plot(time, Wa, xlab = "Time", ylab = "Water absorption (%)")
        f <- as.character(tclObj(f))
        if (f == "Peleg") {
            curve(peleg(x, k1 = 0.1, k2 = 0.9), add = TRUE, col = "red")
        } else if (f == "Logistic") {
            curve(logi(x, a = 70, b1 = 0.1, c1 = 15), add = TRUE, col = "red")
        } else if (f == "Logistic.adapted") {
            curve(logi2(x, a = 70, b = 0.1, c1 = 10, b2 = 0.09, c2 = 50), 
               add = TRUE, col = "red")
        } else if (f == "Peleg.adapted") {
            curve(peleg2(x, k1 = 0.1, k2 = 0.9, b2 = 0.09, c2 = 50), 
               add = TRUE, col = "red")
        }
    }
    OnFit <- function() {
        f <- as.character(tclObj(f))
        a <- as.numeric(tclObj(a))
        k1 <- as.numeric(tclObj(k1))
        k2 <- as.numeric(tclObj(k2))
        b1 <- as.numeric(tclObj(b1))
        b2 <- as.numeric(tclObj(b2))
        c1 <- as.numeric(tclObj(c1))
        c2 <- as.numeric(tclObj(c2))

        if (f == "Peleg") {
           out <- nls(Wa ~ peleg(x = time, k1, k2), 
              start = c(k1 = k1, k2 = k2))
        } else if (f == "Logistic") {
           out <- nls(Wa ~ logi(x = time, a, b1, c1), 
              start = c(a = a, b1 = b1, c1 = c1))
        } else if (f == "Logistic.adapted") {
           out <- nls(Wa ~ logi2(x = time, a, b1, c1, b2, c2), 
              start = c(a = a, b1 = b1, c1 = c1, b1 = b2, c2 = c2))
        } else if (f == "Peleg.adapted") {
           out <- nls(Wa ~ peleg2(x = time, k1, k2, b2, c2), 
              start = c(k1 = k1, k2 = k2, b2 = b2, c2 = c2))
        }
        beta <- coef(out)
        r2 <- round(Rsq(out)$pseudo, 4)
        r2adj <- round(Rsq(out)$adj, 4)
        se <- round(summary(out)$sigma, 6)
        tclvalue(r2Text) <- paste("R-sq: ", as.character(r2))
        tclvalue(r2adjText) <- paste("Adj. R-sq: ", as.character(r2adj))
        tclvalue(seText) <- paste("S.E.: ", as.character(se))

        assign("soakingModels.tmp", out, envir = soakingModels.env)
        plot(time, Wa, xlab = "Time", ylab = "Water absorption (%)")
        if (f == "Peleg") {
            curve(peleg(x, k1 = beta[1], k2 = beta[2]), 
               add = TRUE, col = "blue")
        } else if (f == "Logistic") {
            curve(logi(x, a = beta[1], b1 = beta[2], c1 = beta[3]), 
               add = TRUE, col = "blue")
        } else if (f == "Logistic.adapted") {
            curve(logi2(x, a = beta[1], b1 = beta[2], b2 = beta[3],
               c1 = beta[4], c2 = beta[5]), add = TRUE, col = "blue")
        } else if (f == "Peleg.adapted") {
            curve(peleg2(x, k1 = beta[1], k2 = beta[2], b2 = beta[3],
               c2 = beta[4]), add = TRUE, col = "blue")
        }
    }

    tkpack(frame1, frame2, frame3, frame4, frame5, frame6, 
        frame7, frame8, frame9, fill = "x")
    tkpack(left.frm, right.frm, side = "left", anchor = "n")
    c.but <- tkbutton(base, text = "Clear", command = OnClear)
    q.but <- tkbutton(base, text = "Quit", command = OnQuit)
    fit.but <- tkbutton(base, text = "NLS Estimates", command = OnFit, 
        foreground = "white", background = "navy")
    tkpack(spec.frm)
    tkpack(q.but, side = "right")
    tkpack(c.but, side = "left")
    tkpack(fit.but, side = "right")
    replot()
    tkbind(entry.a, "<Return>", function() replot() )
    tkbind(entry.k1, "<Return>", function() replot() )
    tkbind(entry.k2, "<Return>", function() replot() )
    tkbind(entry.b1, "<Return>", function() replot() )
    tkbind(entry.b2, "<Return>", function() replot() )
    tkbind(entry.c1, "<Return>", function() replot() )
    tkbind(entry.c2, "<Return>", function() replot() )
    tkbind(base, "<Destroy>", function() tclvalue(done) <- 2)
    tkwait.variable(done)
    tkdestroy(base)

    # output
    fit <- get("soakingModels.tmp", envir = soakingModels.env)
    return(fit)
} 


R/dryingmodels.R
dryingmodels <- 
function(moisture, time) 
{
    if (!requireNamespace(c("tcltk", "tkrplot"), quietly = TRUE)) 
        stop("packages tcltk and tkrplot are required")
    geterrmessage()
    done <- tclVar(0)
    dryingModels.env <- new.env()
    assign("dryingModels.tmp", list(), envir = dryingModels.env)
    x <- NULL

    # Moisture ratio
    Mf <- min(moisture)
    Mi <- max(moisture)
    Mr <- (moisture - Mf)/(Mi - Mf)

    # equations
    page <- function(x, K, n) exp(-K*x^n)
    HandP <- function(x, A, K) A*exp(-K*x)
    mHandP <- function(x, A, K, b, K0, K1) A*exp(-K*x) + b*exp(-K0*x) + exp(-K1*x)
    midilli <- function(x, A, K, n, b) A*exp(-K*x^n) + b*x
    diffusion.approx <- function(x, A, K, b) A*exp(-K*x) + (1-A)*exp(-K*b*x)
    two.terms <- function(x, A, b, K0, K1) A*exp(-K0*x) + b*exp(-K1*x)
    two.term.exp <- function(x, A, K) A*exp(-K*x) + (1-A)*exp(-K*A*x)
    logarithmic <- function(x, A, K, b) A*exp(-K*x) + b
    thompson <- function(x, A, b) exp(-A - sqrt(A^2 + 4*b*x)) / (2*b)
    newton <- function(x, K) exp(-K*x)
    verma <- function(x, A, K, K1) A*exp(-K*x) + (1-A)*exp(-K1*x)
    WandS <- function(x, A, b) 1 + A*x + b*x^2

    # input-start variables
    f <- tclVar("Henderson.and.Pabis")
    A <- tclVar("0.9")
    K <- tclVar("0.1")
    n <- tclVar("0.9")
    b <- tclVar("0.1")
    K0 <- tclVar("0.1")
    K1 <- tclVar("0.1")

    replot <- function(...) {
        fun <- as.character(tclObj(f))
        A <- as.numeric(tclObj(A))
        K <- as.numeric(tclObj(K))
        n <- as.numeric(tclObj(n))
        b <- as.numeric(tclObj(b))
        K0 <- as.numeric(tclObj(K0))
        K1 <- as.numeric(tclObj(K1))
        plot(time, Mr, xlab = "Time", ylab = "Moisture ratio")
        if (fun == "Page") {
            curve(page(x, K, n), add = TRUE, col = "red")
        } else if (fun == "Henderson.and.Pabis") {
            curve(HandP(x, A, K), add = TRUE, col = "red")
        } else if (fun == "Modified.Henderson.and.Pabis") {
            curve(mHandP(x, A, K, b, K0, K1), add = TRUE, col = "red")
        } else if (fun == "Midilli") {
            curve(midilli(x, A, K, n, b), add = TRUE, col = "red")
        } else if (fun == "Diffusion.Approximation") {
            curve(diffusion.approx(x, A, K, b), add = TRUE, col = "red")
        } else if (fun == "Two.Terms") {
            curve(two.terms(x, A, b, K0, K1), add = TRUE, col = "red")
        } else if (fun == "Two.Term.Exponential") {
            curve(two.term.exp(x, A, K), add = TRUE, col = "red")
        } else if (fun == "Logarithmic") {
            curve(logarithmic(x, A, K, b), add = TRUE, col = "red")
        } else if (fun == "Thompson") {
            curve(thompson(x, A, b), add = TRUE, col = "red")
        } else if (fun == "Newton") {
            curve(newton(x, K), add = TRUE, col = "red")
        } else if (fun == "Verma") {
            curve(verma(x, A, K, K1), add = TRUE, col = "red")
        } else if (fun == "Wang.and.Singh") {
            curve(WandS(x, A, b), add = TRUE, col = "red")
        }
    }
    redraw <- function(...) {
        fun <- as.character(tclObj(f))
        if (fun == "Page") {
            # state of parameters: A, K, n, b, K0, K1
            states <- c("disabled", "normal", "normal", "disabled", "disabled", "disabled")
            tkfocus(entry.K)
            tkfocus(entry.n)
        } else if (fun == "Henderson.and.Pabis") {
            states <- c("normal", "normal", "disabled", "disabled", "disabled", "disabled")
            tkfocus(entry.A)
            tkfocus(entry.K)
        } else if (fun == "Modified.Henderson.and.Pabis") {
            states <- c("normal", "normal", "disabled", "normal", "normal", "normal")
            tkfocus(entry.A)
            tkfocus(entry.K)
            tkfocus(entry.b)
            tkfocus(entry.K0)
            tkfocus(entry.K1)
        } else if (fun == "Midilli") {
            states <- c("normal", "normal", "normal", "normal", "disabled", "disabled")
            tkfocus(entry.A)
            tkfocus(entry.K)
            tkfocus(entry.n)
            tkfocus(entry.b)
        } else if (fun == "Diffusion.Approximation") {
            states <- c("normal", "normal", "disabled", "normal", "disabled", "disabled")
            tkfocus(entry.A)
            tkfocus(entry.K)
            tkfocus(entry.b)
        } else if (fun == "Two.Terms") {
            states <- c("normal", "disabled", "disabled", "normal", "normal", "normal")
            tkfocus(entry.A)
            tkfocus(entry.b)
            tkfocus(entry.K0)
            tkfocus(entry.K1)
        } else if (fun == "Two.Term.Exponential") {
            states <- c("normal", "normal", "disabled", "disabled", "disabled", "disabled")
            tkfocus(entry.A)
            tkfocus(entry.K)
        } else if (fun == "Logarithmic") {
            states <- c("normal", "normal", "disabled", "normal", "disabled", "disabled")
            tkfocus(entry.A)
            tkfocus(entry.K)
            tkfocus(entry.b)
        } else if (fun == "Thompson") {
            states <- c("normal", "disabled", "disabled", "normal", "disabled", "disabled")
            tkfocus(entry.A)
            tkfocus(entry.b)
        } else if (fun == "Newton") {
            states <- c("disabled", "normal", "disabled", "disabled", "disabled", "disabled")
            tkfocus(entry.K)
        } else if (fun == "Verma") {
            states <- c("normal", "normal", "disabled", "disabled", "disabled", "normal")
            tkfocus(entry.A)
            tkfocus(entry.K)
            tkfocus(entry.K1)
        } else if (fun == "Wang.and.Singh") {
            states <- c("normal", "disabled", "disabled", "normal", "disabled", "disabled")
            tkfocus(entry.A)
            tkfocus(entry.b)
        }
        tkconfigure(entry.A, state = states[1])
        tkconfigure(ts1, state = states[1]) 
        tkconfigure(entry.K, state = states[2])
        tkconfigure(ts3, state = states[2]) 
        tkconfigure(entry.n, state = states[3])
        tkconfigure(ts4, state = states[3])  
        tkconfigure(entry.b, state = states[4])
        tkconfigure(ts6, state = states[4])  
        tkconfigure(entry.K0, state = states[5])
        tkconfigure(ts7, state = states[5])  
        tkconfigure(entry.K1, state = states[6])
        tkconfigure(ts8, state = states[6])  
        replot()
    }
    base <- tktoplevel()
    tkwm.title(base, "Seed Drying Kinetics")
    spec.frm <- tkframe(base, borderwidth = 2)
    left.frm <- tkframe(spec.frm)
    right.frm <- tkframe(spec.frm)  

    # parameter: A
    frame1 <- tkframe(left.frm, relief = "groove", borderwidth = 2)
    tkpack(tklabel(frame1, text = "Parameters"), fill = "both", side = "top")
    entry.A <- tkentry(frame1, width = "5", textvariable = A)
    tkpack(ts1 <- tkscale(frame1, label = "A", command = replot, 
        from = -0.5, to = 5, showvalue = 0, variable = A, 
        resolution = 0.001, orient = "horiz", relief = "groove"), 
        fill = "both", expand = 1, padx = 3, pady = 2, ipadx = 3, 
        ipady = 2, side = "left")
    tkpack(entry.A, fill = "none", side = "right")

    # models
    frame2 <- tkframe(right.frm, relief = "groove", borderwidth = 2)
    tkpack(tklabel(frame2, text = "Models"))
    for (i in c("Page", "Henderson.and.Pabis", 
        "Modified.Henderson.and.Pabis", "Midilli", "Diffusion.Approximation", 
        "Two.Terms", "Two.Term.Exponential", "Logarithmic", "Thompson",
        "Newton", "Verma", "Wang.and.Singh")) {
        tmp <- tkradiobutton(frame2, command = redraw, text = i, 
            value = i, variable = f)
        tkpack(tmp, anchor = "w")
    }

    # parameter: K
    frame3 <- tkframe(left.frm, relief = "groove", borderwidth = 2)
    entry.K <- tkentry(frame3, width = "5", textvariable = K)
    tkpack(ts3 <- tkscale(frame3, label = "K", command = replot, 
        from = 0, to = 5, showvalue = 0, variable = K,
        resolution = 0.01, orient = "horiz", relief = "groove"), 
        fill = "both", expand = 1, padx = 3, pady = 2, ipadx = 3, 
        ipady = 2, side = "left")
    tkpack(entry.K, fill = "none", side = "right")

    # parameter: n
    frame4 <- tkframe(left.frm, relief = "groove", borderwidth = 2)
    entry.n <- tkentry(frame4, width = "5", textvariable = n)
    tkpack(ts4 <- tkscale(frame4, label = "n", 
        command = replot, from = 0, to = 1.5, showvalue = 0, 
        variable = n, resolution = 0.01, orient = "horiz", 
        relief = "groove"), fill = "none", expand = 1, padx = 3, 
        pady = 2, ipadx = 3, ipady = 2, side = "left")
    tkpack(entry.n, side = "right")

    # fitting criteria
    frame5 <- tkframe(right.frm, relief = "groove", borderwidth = 2)
    r2Text <- tclVar("R-sq: ")
    label.r2 <- tklabel(frame5, text=tclvalue(r2Text))
    tkconfigure(label.r2, textvariable=r2Text)
    tkgrid(label.r2)
    r2adjText <- tclVar("Adj. R-sq: ")
    label.r2adj <- tklabel(frame5, text=tclvalue(r2adjText))
    tkconfigure(label.r2adj, textvariable=r2adjText)
    tkgrid(label.r2adj)
    seText <- tclVar("S.E.: ")
    label.se <- tklabel(frame5, text=tclvalue(seText))
    tkconfigure(label.se, textvariable=seText)
    tkgrid(label.se)

    # parameter: b
    frame6 <- tkframe(left.frm, relief = "groove", borderwidth = 2)
    entry.b <- tkentry(frame6, width = "5", textvariable = b)
    tkpack(ts6 <- tkscale(frame6, label = "b", 
        command = replot, from = -1, to = 2, showvalue = 0, 
        variable = b, resolution = 0.001, orient = "horiz", 
        relief = "groove"), fill = "none", expand = 1, padx = 3, 
        pady = 2, ipadx = 3, ipady = 2, side = "left")
    tkpack(entry.b, side = "right")

    # parameter: K0
    frame7 <- tkframe(left.frm, relief = "groove", borderwidth = 2)
    entry.K0 <- tkentry(frame7, width = "5", textvariable = K0)
    tkpack(ts7 <- tkscale(frame7, label = "K0", 
        command = replot, from = -1, to = 1.5, showvalue = 0, 
        variable = K0, resolution = 0.01, orient = "horiz", 
        relief = "groove"), fill = "none", expand = 1, padx = 3, 
        pady = 2, ipadx = 3, ipady = 2, side = "left")
    tkpack(entry.K0, side = "right")

    # parameter: K1
    frame8 <- tkframe(left.frm, relief = "groove", borderwidth = 2)
    entry.K1 <- tkentry(frame8, width = "5", textvariable = K1)
    tkpack(ts8 <- tkscale(frame8, label = "K1", 
        command = replot, from = -1, to = 1.5, showvalue = 0, 
        variable = K1, resolution = 0.01, orient = "horiz", 
        relief = "groove"), fill = "none", expand = 1, padx = 3, 
        pady = 2, ipadx = 3, ipady = 2, side = "left")
    tkpack(entry.K1, side = "right")

    # buttons
    OnQuit <- function() {
        tclvalue(done) <- 2
    }
    OnClear <- function() {
        assign("dryingModels.tmp", list(), envir = dryingModels.env)
        plot(time, Mr, xlab = "Time", ylab = "Moisture ratio")
        f <- as.character(tclObj(f))
        if (f == "Page") {
            curve(page(x, K = 0.1, n = 0.7), add = TRUE, col = "red")
        } else if (f == "Henderson.and.Page") {
            curve(HandP(x, A = 1, K = 0.12), add = TRUE, col = "red")
        } else if (f == "Modified.Henderson.and.Page") {
            curve(mHandP(x, A = 0.9, K = 0.1, b = 0.1, K0 = 0.1, K1 = 0.1), 
               add = TRUE, col = "red")
        } else if (f == "Midilli") {
            curve(midilli(x, A = 0.9, K = 0.1, n = 0.7, b = 0.01), 
               add = TRUE, col = "red")
        } else if (f == "Diffusion.Approximation") {
            curve(diffusion.approx(x, A = 0.9, K = 0.1, b = 0.1), 
               add = TRUE, col = "red")
        } else if (f == "Two.Terms") {
            curve(two.terms(x, A = 0.9, b = 0.1, K0 = 0.1, K1 = 0.1), 
               add = TRUE, col = "red")
        } else if (f == "Two.Term.Exponential") {
            curve(two.term.exp(x, A = 1.5, K = 0.2), 
               add = TRUE, col = "red")
        } else if (f == "Logarithmic") {
            curve(logarithmic(x, A = 0.9, K = 0.1, b = 0.1), 
               add = TRUE, col = "red")
        } else if (f == "Thompson") {
            curve(thompson(x, A = 0.9, b = 0.1), add = TRUE, col = "red")
        } else if (f == "Newton") {
            curve(newton(x, K = 0.1), add = TRUE, col = "red")
        } else if (f == "Verma") {
            curve(verma(x, A = 0.1, K = 5, K1 = 0.1),
               add = TRUE, col = "red")
        } else if (f == "Wang.and.Singh") {
            curve(WandS(x, A = 0.01, b = 0.001),
               add = TRUE, col = "red")
        }
    }
    OnFit <- function() {
        f <- as.character(tclObj(f))
        A <- as.numeric(tclObj(A))
        K <- as.numeric(tclObj(K))
        n <- as.numeric(tclObj(n))
        b <- as.numeric(tclObj(b))
        K0 <- as.numeric(tclObj(K0))
        K1 <- as.numeric(tclObj(K1))
        if (f == "Page") {
           out <- nls(Mr ~ page(x = time, K, n), 
              start = c(K = K, n = n))
        } else if (f == "Henderson.and.Pabis") {
           out <- nls(Mr ~ HandP(x = time, A, K), 
              start = c(A = A, K = K))
        } else if (f == "Modified.Henderson.and.Page") {
           out <- nls(Mr ~ mHandP(x = time, A, K, b, K0, K1), 
              start = c(A = A, K = K, b = b, K0 = K0, K1 = K1))
        } else if (f == "Midilli") {
           out <- nls(Mr ~ midilli(x = time, A, K, n, b), 
              start = c(A = A, K = K, n = n, b = b))
        } else if (f == "Diffusion.Approximation") {
           out <- nls(Mr ~ diffusion.approx(x = time, A, K, b), 
              start = c(A = A, K = K, b = b))
        } else if (f == "Two.Terms") {
           out <- nls(Mr ~ two.terms(x = time, A, b, K0, K1), 
              start = c(A = A, b = b, K0 = K0, K1 = K1))
        } else if (f == "Two.Term.Exponential") {
           out <- nls(Mr ~ two.term.exp(x = time, A, K), 
              start = c(A = A, K = K))
        } else if (f == "Logarithmic") {
           out <- nls(Mr ~ logarithmic(x = time, A, K, b), 
              start = c(A = A, K = K, b = b))
        } else if (f == "Thompson") {
           out <- nls(Mr ~ thompson(x = time, A, b), 
              start = c(A = A, b = b))
        } else if (f == "Newton") {
           out <- nls(Mr ~ newton(x = time, K), start = c(K = K))
        } else if (f == "Verma") {
           out <- nls(Mr ~ verma(x = time, A, K, K1), 
              start = c(A = A, K = K, K1 = K1))
        } else if (f == "Wang.and.Singh") {
           out <- nls(Mr ~ WandS(x = time, A, b), 
              start = c(A = A, b = b))
        }
        beta <- coef(out)
        r2 <- round(Rsq(out)$pseudo, 4)
        r2adj <- round(Rsq(out)$adj, 4)
        se <- round(summary(out)$sigma, 6)
        tclvalue(r2Text) <- paste("R-sq: ", as.character(r2))
        tclvalue(r2adjText) <- paste("Adj. R-sq: ", as.character(r2adj))
        tclvalue(seText) <- paste("S.E.: ", as.character(se))

        assign("dryingModels.tmp", out, envir = dryingModels.env)
        plot(time, Mr, xlab = "Time", ylab = "Moisture ratio")
        if (f == "Page") {
            curve(page(x, K = beta[1], n = beta[2]), 
               add = TRUE, col = "blue")
        } else if (f == "Henderson.and.Pabis") {
            curve(HandP(x, A = beta[1], K = beta[2]), 
               add = TRUE, col = "blue")
        } else if (f == "Modified.Henderson.and.Page") {
            curve(mHandP(x, A = beta[1], K = beta[2], b = beta[3],
               K0 = beta[4], K1 = beta[5]), add = TRUE, col = "blue")
        } else if (f == "Midilli") {
            curve(midilli(x, A = beta[1], K = beta[2], n = beta[3],
               b = beta[4]), add = TRUE, col = "blue")
        } else if (f == "Diffusion.Approximation") {
            curve(diffusion.approx(x, A = beta[1], K = beta[2], 
               b = beta[3]), add = TRUE, col = "blue")
        } else if (f == "Two.Terms") {
            curve(two.terms(x, A = beta[1], b = beta[2], K0 = beta[3], 
               K1 = beta[4]), add = TRUE, col = "blue")
        } else if (f == "Two.Term.Exponential") {
            curve(two.term.exp(x, A = beta[1], K = beta[2]), 
               add = TRUE, col = "blue")
        } else if (f == "Logarithmic") {
            curve(logarithmic(x, A = beta[1], K = beta[2], b = beta[3]), 
               add = TRUE, col = "blue")
        } else if (f == "Thompson") {
            curve(thompson(x, A = beta[1], b = beta[2]), 
               add = TRUE, col = "blue")
        } else if (f == "Newton") {
            curve(newton(x, K = beta[1]), add = TRUE, col = "blue")
        } else if (f == "Verma") {
            curve(verma(x, A = beta[1], K = beta[2], K1 = beta[3]), 
               add = TRUE, col = "blue")
        } else if (f == "Wang.and.Singh") {
            curve(WandS(x, A = beta[1], b = beta[2]), 
               add = TRUE, col = "blue")
        }
    }

    tkpack(frame1, frame2, frame3, frame4, frame5, frame6, 
        frame7, frame8, fill = "x")
    tkpack(left.frm, right.frm, side = "left", anchor = "n")
    c.but <- tkbutton(base, text = "Clear", command = OnClear)
    q.but <- tkbutton(base, text = "Quit", command = OnQuit)
    fit.but <- tkbutton(base, text = "NLS Estimates", command = OnFit, 
        foreground = "white", background = "navy")
    tkpack(spec.frm)
    tkpack(q.but, side = "right")
    tkpack(c.but, side = "left")
    tkpack(fit.but, side = "right")
    replot()
    tkbind(entry.A, "<Return>", function() replot() )
    tkbind(entry.K, "<Return>", function() replot() )
    tkbind(entry.n, "<Return>", function() replot() )
    tkbind(entry.b, "<Return>", function() replot() )
    tkbind(entry.K0, "<Return>", function() replot() )
    tkbind(entry.K1, "<Return>", function() replot() )
    tkbind(base, "<Destroy>", function() tclvalue(done) <- 2)
    tkwait.variable(done)
    tkdestroy(base)

    # output
    fit <- get("dryingModels.tmp", envir = dryingModels.env)
    return(fit)
} 
